src/sinks/datadog/logs/sink.rs

```diff
@@ -158,96 +129,168 @@ impl From<io::Error> for RequestBuildError {
     }
 }
 
+impl From<serde_json::Error> for RequestBuildError {
+    fn from(error: serde_json::Error) -> RequestBuildError {
+        RequestBuildError::Json { error }
+    }
+}
+
 struct LogRequestBuilder {
     default_api_key: Arc<str>,
-    encoding: JsonEncoding,
+    transformer: Transformer,
     compression: Compression,
 }
 
-impl RequestBuilder<(Option<Arc<str>>, Vec<Event>)> for LogRequestBuilder {
-    type Metadata = (Arc<str>, EventFinalizers);
-    type Events = Vec<Event>;
-    type Encoder = JsonEncoding;
-    type Payload = Bytes;
-    type Request = LogApiRequest;
-    type Error = RequestBuildError;
+struct CountingWrite<'a, T> {
+    inner: T,
+    count: &'a std::cell::Cell<usize>,
+}
+
+impl<'a, T> CountingWrite<'a, T> {
+    fn into_inner(self) -> T {
+        self.inner
+    }
+}
 
-    fn compression(&self) -> Compression {
-        self.compression
+#[allow(clippy::disallowed_methods)]
+impl<'a, T: std::io::Write> std::io::Write for CountingWrite<'a, T> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.count.set(self.count.get() + buf.len());
+        self.inner.write(buf)
     }
 
-    fn encoder(&self) -> &Self::Encoder {
-        &self.encoding
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
     }
+}
 
-    fn split_input(
+impl LogRequestBuilder {
+    fn build_request(
```

Comment:
I think the logic here is more complex than it needs to be due to the limitations of `serde_json::Serializer::serialize_seq` in `try_serialize`. Specifically, the fact that we cannot remove a serialized entry if we go over the size limit causes us to have to track more state and repeat work. Rather, if we manually build the json array payload, then the logic can be greatly simplified.

We could replace `try_serialize` with the following function.
```
/// Serialize events into a buffer as a JSON array that has a maximum size of
/// `MAX_PAYLOAD_BYTES`.
///
/// Returns the serialized events, the buffer, and the byte size of the events.
/// Events that are not serialized remain in the `events` parameter.
fn serialize_with_capacity(
    events: &mut VecDeque<(Event, JsonSize)>,
) -> Result<(Vec<Event>, Vec<u8>, GroupedCountByteSize), io::Error> {
    // Compute estimated size, accounting for the size of the brackets and commas.
    let total_estimated =
        events.iter().map(|(_, size)| size.get()).sum::<usize>() + events.len() * 2;

    // Initialize state.
    let mut buf = Vec::with_capacity(total_estimated);
    let mut byte_size = telemetry().create_request_count_byte_size();
    let mut events_serialized = Vec::with_capacity(events.len());

    // Write entries until the buffer is full.
    buf.write(b"[")?;
    let mut first = true;
    while let Some((event, estimated_json_size)) = events.pop_front() {
        // Track the existing length of the buffer so we can truncate it if we need to.
        let existing_len = buf.len();
        if first {
            first = false;
        } else {
            buf.write(b",")?;
        }
        serde_json::to_writer(&mut buf, event.as_log())?;
        // If the buffer is too big, truncate it and break out of the loop.
        if buf.len() >= MAX_PAYLOAD_BYTES {
            events.push_front((event, estimated_json_size));
            buf.truncate(existing_len);
            break;
        }
        // Otherwise, track the size of the event and continue.
        byte_size.add_event(&event, estimated_json_size);
        events_serialized.push(event);
    }
    buf.write(b"]")?;

    Ok((events_serialized, buf, byte_size))
}
```

And then `build_request` would become the following:
```
    fn build_request(
        &self,
        mut events: Vec<Event>,
        api_key: Arc<str>,
    ) -> Result<Vec<LogApiRequest>, RequestBuildError> {
        // Transform events and pre-compute their estimated size.
        let mut events_with_estimated_size: VecDeque<(Event, JsonSize)> = events
            .iter_mut()
            .map(|event| {
                normalize_event(event);
                self.transformer.transform(event);
                let estimated_json_size = event.estimated_json_encoded_size_of();
                (event.clone(), estimated_json_size)
            })
            .collect();

        // Construct requests respecting the max payload size.
        let mut requests: Vec<LogApiRequest> = Vec::new();
        while !events_with_estimated_size.is_empty() {
            let (events_serialized, body, byte_size) =
                serialize_with_capacity(&mut events_with_estimated_size)?;
            if events_serialized.is_empty() {
                // first event was too large for whole request
                let _too_big = events_with_estimated_size.pop_front();
                emit!(ComponentEventsDropped::<UNINTENTIONAL> {
                    count: 1,
                    reason: "Event too large to encode."
                });
            } else {
                let request =
                    self.finish_request(body, events_serialized, byte_size, Arc::clone(&api_key))?;
                requests.push(request);
            }
        }

        Ok(requests)
    }
```

What do you think? I can commit my working branch if you agree with this approach.

Reply:
I like it! This seems like a good logical next step to improve it. Thanks for that, as I was running out of gas on this change a little bit 😆